[FSDP] Rework meta device init

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #104189

This addresses https://github.com/pytorch/pytorch/issues/104187.

After this PR, the contract with the user is that:
- If passing `param_init_fn=None`, each `nn.Module.reset_parameters()` should only initialize its own parameters/buffers (like `parameters(recurse=False)`/`buffers(recurse=False)`).
- If passing `param_init_fn` not equal to `None`, then similarly, one call to `param_init_fn(module)` should only initialize `module`'s own parameters/buffers.

With this contract and this PR's changes, meta device initialization through either `reset_parameters()` or `param_init_fn` should be correct. Those functions will run on the original parameter/buffer shapes allowing for correct shape-dependent computations like for fan-in/fan-out, and there will not be any re-initialization of any module.

Commit:
[FSDP] Rework meta device init

[ghstack-poisoned]

Commit:
Update on "[FSDP] Rework meta device init"


This addresses https://github.com/pytorch/pytorch/issues/104187.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Rework meta device init"


This addresses https://github.com/pytorch/pytorch/issues/104187.

After this PR, the contract with the user is that:
- If passing `param_init_fn=None`, each `nn.Module.reset_parameters()` should only initialize its own parameters/buffers (like `parameters(recurse=False)`/`buffers(recurse=False)`).
- If passing `param_init_fn` not equal to `None`, then similarly, one call to `param_init_fn(module)` should only initialize `module`'s own parameters/buffers.

With this contract and this PR's changes, meta device initialization through either `reset_parameters()` or `param_init_fn` should be correct. Those functions will run on the original parameter/buffer shapes allowing for correct shape-dependent computations like for fan-in/fan-out, and there will not be any re-initialization of any module.


[ghstack-poisoned]

rohan-varma commented on test/distributed/fsdp/test_fsdp_meta.py:
```diff
@@ -43,11 +43,15 @@
+            # Assume that a module has `reset_parameters()` iff it has directly
+            # managed parameters or buffers
+            if hasattr(module, "reset_parameters"):
+                module.reset_parameters()
```
seems that the contract is now that reset_parameter is nonrecursive, did we document this somewhere?

awgu replied:
Yep, see the changes to the public docstrings in fully_sharded_data_parallel.py.

rohan-varma commented on test/distributed/fsdp/test_fsdp_meta.py:
```diff
@@ -59,24 +63,33 @@ def __init__(self, *args, **kwargs):
+
+    def reset_parameters(self, *args, **kwargs):
+        torch.manual_seed(42)
+        # Use an initialization method that depends on shape
```
curious, what does this change with regard to the test?

awgu replied:
If we fill with 1 always, then it does not matter if we accidentally re-initialize a nested module that has already been sharded. Here, because the initialization function depends on the parameter shape, the test would fail correctness if we accidentally re-initialized when the parameter had the sharded shape.

rohan-varma commented on test/distributed/fsdp/test_fsdp_meta.py:
```diff
@@ -90,25 +103,22 @@ def __init__(self, device):
-    if is_meta:
-        module.to_empty(device=torch.cuda.current_device())
-    with torch.no_grad():
+    has_meta_states = any(p.is_meta for p in module.parameters(recurse=False)) or any(
```
nit:

```
has_meta_states = any(tensor.is_meta for tensor in itertools.chain(module.parameters, module.buffers))
```

rohan-varma commented on test/distributed/fsdp/test_fsdp_meta.py:
```diff
@@ -90,25 +103,22 @@ def __init__(self, device):
+    has_meta_states = any(p.is_meta for p in module.parameters(recurse=False)) or any(
+        b.is_meta for b in module.buffers(recurse=False)
+    )
+    device = torch.device("cuda", torch.cuda.current_device())
```
define device after L116, otherwise its not used?

rohan-varma commented on test/distributed/fsdp/test_fsdp_meta.py:
```diff
@@ -225,7 +235,9 @@ def _test_nested_model_with_meta_device(
             module = meta_module_fn()
-            is_meta = next(module.parameters()).is_meta
+            is_meta = (
+                next(module.parameters()).is_meta or next(module.buffers()).is_meta
```
thanks for adding the buffers! would a test with only buffers being meta have picked this issue up? if so, should we add it?

awgu replied:
Yes, I had to add this since I added a buffer-only module.

rohan-varma commented on torch/distributed/fsdp/_init_utils.py:
```diff
@@ -488,7 +488,7 @@ def _init_param_handle_from_module(
     device_from_device_id = _get_device_from_device_id(device_id, state.rank)
     is_meta_module, is_torchdistX_deferred_init = _need_to_materialize_module(
-        fully_sharded_module, state._ignored_params
+        fully_sharded_module, state._ignored_params, state._ignored_modules
```
this looks like a former bug too, did we add unittests here?

awgu replied:
We are just using ignored modules to skip their buffers. Otherwise, normally `_ignored_params` is derived from `_ignored_modules` (except for the parameter `ignored_states` path, which is a known bug).

rohan-varma commented on torch/distributed/fsdp/_init_utils.py:
```diff
@@ -838,6 +839,14 @@ def _need_to_materialize_module(
     managed_params = list(_get_orig_params(module, ignored_params))
     is_meta_module = any(param.is_meta for param in managed_params)
+    # TODO: We need to establish a contract for FSDP and buffers. For now, we
+    # skip checking for meta buffers from ignored modules. We should consider
```
what about meta parameters from ignored modules?

awgu replied:
Yes, we need to discuss that too in terms of FSDP's contract. For now, meta parameters from ignored modules are also not considered (i.e. they will not make the module considered to need materialization).

rohan-varma commented on torch/distributed/fsdp/_init_utils.py:
```diff
@@ -847,37 +856,69 @@ def _need_to_materialize_module(
+            for module in modules_to_materialize:
+                # As a contract to the user, only call `reset_parameters()` if
+                # the module has directly managed parameters/buffers
+                has_module_states = (
```
maybe
```
has_module_states = len(list(itertools.chain(module.parameters, modeul.buffers)))
```

rohan-varma commented on torch/distributed/fsdp/_init_utils.py:
```diff
@@ -847,37 +856,69 @@ def _need_to_materialize_module(
-            f"device with error {str(e)}. Please ensure your "
-            "module implements a `reset_parameters()` method."
+            f"device with error {str(e)}. Please ensure that your module of"
+            f"type {type(module)} implements a `reset_parameters()` method."
```
this BaseException was before this PR, but shouldn't we be more specific and except a NotImplementedError? Seems like this will catch all exceptions and maybe report a misleading warning if there was some other issue

awgu replied:
I will look into this in a follow-up.

rohan-varma commented on torch/distributed/fsdp/_init_utils.py:
```diff
@@ -847,37 +856,69 @@ def _need_to_materialize_module(
+                child_module not in visited_modules
+                and _get_module_fsdp_state(child_module) is None
+            ):
+                visited_modules.add(child_module)
```
curious, doesn't bfs usually mark the node as visited when it is being processed / popped from the queue?  not sure if that makes a difference

awgu replied:
Haha I don't think it makes a difference. This way is technically better for general graphs since you are marking a node visited as early as possible (otherwise, a node in the queue before this child node may try to visit the child node again).

rohan-varma reviewed (approved):
LGTM, some minor q

Commit:
Update on "[FSDP] Rework meta device init"


This addresses https://github.com/pytorch/pytorch/issues/104187.

After this PR, the contract with the user is that:
- If passing `param_init_fn=None`, each `nn.Module.reset_parameters()` should only initialize its own parameters/buffers (like `parameters(recurse=False)`/`buffers(recurse=False)`).
- If passing `param_init_fn` not equal to `None`, then similarly, one call to `param_init_fn(module)` should only initialize `module`'s own parameters/buffers.

With this contract and this PR's changes, meta device initialization through either `reset_parameters()` or `param_init_fn` should be correct. Those functions will run on the original parameter/buffer shapes allowing for correct shape-dependent computations like for fan-in/fan-out, and there will not be any re-initialization of any module.


[ghstack-poisoned]

awgu commented:
@pytorchbot merge
